File: triton-soccer-ai/src/main/java/com/triton/networking/UDP_Client.java
package com.triton.networking;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.net.*;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.function.Consumer;

public class UDP_Client extends Thread {
    private static final int BUF_SIZE = 9999;
    public final int serverPort;
    private final InetAddress serverAddress;
    private final Consumer<byte[]> callbackPacket;

    private final DatagramSocket socket;
    private final BlockingQueue<byte[]> sendQueue;

    public UDP_Client(String serverAddress, int serverPort, Consumer<byte[]> callbackPacket) throws UnknownHostException, SocketException {
        this(serverAddress, serverPort, callbackPacket, 100);
    }

    public UDP_Client(String serverAddress, int serverPort, Consumer<byte[]> callbackPacket, int timeout) throws UnknownHostException, SocketException {
        super();
        this.serverAddress = InetAddress.getByName(serverAddress);
        this.serverPort = serverPort;
        this.callbackPacket = callbackPacket;

        socket = new DatagramSocket();
        socket.setSoTimeout(timeout);
        sendQueue = new LinkedBlockingQueue<>(5);
    }

    @Override
    public void run() {
        while (!isInterrupted())
            receive(send());
    }

    private void receive(boolean receive) {
        if (!receive || callbackPacket == null) return;

        byte[] buf = new byte[BUF_SIZE];
        DatagramPacket packet = new DatagramPacket(buf, buf.length);
        try {
            socket.receive(packet);
        } catch (IOException e) {
            return;
        }

        try {
            ByteArrayInputStream stream = new ByteArrayInputStream(packet.getData(),
                    packet.getOffset(),
                    packet.getLength());
            byte[] bytes = stream.readAllBytes();
            stream.close();
            callbackPacket.accept(bytes);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private boolean send() {
        try {
            byte[] bytes = sendQueue.take();
            DatagramPacket packet = new DatagramPacket(bytes, bytes.length, serverAddress, serverPort);
            socket.send(packet);
        } catch (InterruptedException | IOException e) {
            e.printStackTrace();
        }
        return true;
    }

    public void addSend(byte[] bytes) {
        if (sendQueue.remainingCapacity() == 0)
            sendQueue.poll();
        sendQueue.add(bytes);
    }
}

File: triton-bot/src/main/python/networking/udp_client.py
from os import wait
import socket
from threading import Thread
from queue import Empty, Queue


class UDP_Client(Thread):
    BUF_SIZE = 9999
    QUEUE_CAPACITY = 5

    def __init__(self, server_address, server_port, callback, timeout=0):
        super().__init__()
        self.server_address = server_address
        self.server_port = server_port
        self.callback = callback

        self.socket = socket.socket(family=socket.AF_INET, type=socket.SOCK_DGRAM)
        self.socket.settimeout(timeout)
        self.send_queue = Queue(5)

    def run(self):
        super().run()
        while (True):
            self.receive(self.send())

    def send(self):
        try:
            bytes = self.send_queue.get()
        except Empty:
            return False
        
        self.socket.sendto(bytes, (self.server_address, self.server_port))
        return True

    def receive(self, receive):
        if (not receive or self.callback == None):
            return

        try:
            packet = self.socket.recvfrom(UDP_Client.BUF_SIZE)
        except socket.timeout:
            return
        except socket.error:
            return

        self.callback(bytes=packet[0])

    def add_send(self, bytes):
        if (self.send_queue.full()):
            self.send_queue.get()
        self.send_queue.put(bytes)